Jellyfin.Server.Implementations/UserData/UserDataManager.cs

```diff
@@ -0,0 +1,262 @@
+    /// </exception>
+    /// <returns>A <see cref="Task"/> to save the userdata against an item.</returns>
+    public async Task SaveUserDataAsync(User? user, BaseItem item, UserItemData userData, UserDataSaveReason reason, CancellationToken cancellationToken)
+    {
+        ArgumentNullException.ThrowIfNull(userData);
+
+        ArgumentNullException.ThrowIfNull(item);
+
+        ArgumentNullException.ThrowIfNull(user);
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        var keys = item.GetUserDataKeys();
+
+        var dbContext = await _provider.CreateDbContextAsync(cancellationToken).ConfigureAwait(false);
+        await using (dbContext.ConfigureAwait(false))
+        {
+            foreach (var key in keys)
+            {
+                userData.Key = key;
+                dbContext.UserDatas.Update(userData);
+            }
+
+            await dbContext.SaveChangesAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        UserDataSaved?.Invoke(this, new UserDataSaveEventArgs
+        {
+            Keys = keys,
+            UserData = userData,
+            SaveReason = reason,
+            UserId = user!.Id,
+            Item = item
+        });
+    }
+
+    public void SaveUserData(Guid userId, BaseItem item, UserItemData userData, UserDataSaveReason reason, CancellationToken cancellationToken)
+    {
+        var user = _userManager.GetUserById(userId);
+        SaveUserData(user, item, userData, reason, cancellationToken);
+    }
+
+    public void SaveUserData(User? user, BaseItem item, UserItemData userData, UserDataSaveReason reason, CancellationToken cancellationToken)
+    {
+        SaveUserDataAsync(user, item, userData, reason, cancellationToken).ConfigureAwait(false);
+    }
+
+    private async Task<UserItemData> GetUserDataAsync(User? user, BaseItem item)
+    {
+        ArgumentNullException.ThrowIfNull(user);
+        ArgumentNullException.ThrowIfNull(item);
+        var dbContext = await _provider.CreateDbContextAsync().ConfigureAwait(false);
+        UserItemData? userItemData;
+        await using (dbContext.ConfigureAwait(false))
+        {
+            userItemData = await dbContext.UserDatas.FindAsync(new object[] { user.Id, item.Id.ToString() }).ConfigureAwait(true);
+            if (userItemData == null)
+            {
+                userItemData = new UserItemData
+                {
+                    Key = item.Id.ToString(),
+                    UserId = user.Id
+                };
+            }
+        }
+
+        return userItemData;
+    }
+
+    [Obsolete("Use GetUserDataAsync instead.")]
+    public UserItemData GetUserData(User? user, BaseItem item)
+    {
+        return GetUserDataAsync(user, item).GetAwaiter().GetResult();
+    }
+
+    public UserItemDataDto GetUserDataDto(BaseItem item, User user)
+    {
+             var userData = GetUserDataAsync(user, item).GetAwaiter().GetResult();
```

Comment:
Same here

Reply:
> There are 20+ usages of GetUserDataDto and they are all in non-async code.

Then make then async

Reply:
My issue with that is we get this PR in and the follow up one never gets though and we have another mess in the code where we supposed to use cleaned it up.

Then just make the whole stack sync and reintroduce the async version later.

Reply:
I'd say EF migration is the main point, everything else can be tackled later.

Currently we are bound to SQLite anyways which is mostly sychronous, so switching everything to async would not provide much right now. In the long run it is still something we should adopt to, but I don't think this is the scope of this PR.

Reply:
@JPVenson brings up good points, but this PR is already massive, and is only going to become larger before getting merged into master.

With touching almost 100 files, and with about 3500 lines to review, it is already a lot of work keeping on top of merge conflicts.

It would be best to keep the files touched minimal, even at some detriment of overall code quality and performance, in order to make it easier to deal with this massive task, that we've been postponing for many years already due to the amount of work that «needs to happen at once».

The requested change sound like would be relatively easy to do standalone, in one or more PRs down the line.
As such, I'm marking this conversation as resolved.

Consider making a cleanup issue with details on the cleanup task, pointing back to this conversation if you think that is desirable. Could possibly be a «good first issue»?